rfcs/convergence/root-as-slot.md

```diff
@@ -0,0 +1,248 @@
+# RFC: root as a slot
+
+---
+
+_@bsunderhus @layershifter @ling1726_
+
+## Summary
+
+This RFC proposes to treat `root` as a _slot_ to simplify signature, improve `Typings` and avoid conflicts between `root` and other _slots_ signatures.
+
+Main changes would be:
+
+1. `root` becomes a regular slot on component state
+2. Support references on every _slot_
+3. Supports Custom Elements on every _slot_
+4. Allows a restricted `as` prop on every slot
+5. Stops using `getNativeElementProps` on every slot, will be used only for `root`
+
+## Problem statement
+
+There are disparities in how `root` is treated compared to _slots_. Some discussions have popped up due to these divergences between _slots_ and `root`:
+
+1. [Do not support as prop for slots that render native DOM elements] conflicts with [1st rule of ARIA opt-out mechanism]
+2. Support for Custom Elements on _slots_ but not for _root_
+3. Right now `Typings` for `root` aren't available as there's no way to split `root` from internal state.
+4. There's no possible way of passing refs to slots but `root` does
+5. [Primary slot]
+
+Major differences between `root` (shown below as 🌿) and _slots_ (shown below as 🎰):
+
+1. References 🌿
+2. Custom Elements as base element 🎰
+3. `as` prop 🌿
+4. Well defined Types 🎰
+
+## Detailed Design or Proposal
+
+For the user of the component perspective there will be no changes in the signature.
+
+### Converge `root` to a _slot_
+
+This RFC proposes to treat `root` as a _slot_.
+
+Treating `root` as a slot means having to declare `root` _Typings_ together with other _slots_. This will impact in a components state signature, where `root` will have to be declared.
+
+> ⚠️ That doesn't mean `root` will be available in component's properties interface
+
+#### Before
+
+```ts
+export type Slots = {
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> & React.HTMLAttributes<HTMLElement> {/* ... */}
+export interface State extends ComponentState<Slots> {/* ... */}
+
+// use*State hook
+function useState({ slot, ...props }: Props): State {
+  // ...
+  return {
+    ...props,
+    slot: resolveShorthand(slot),
+  };
+}
+```
+
+With the modifications:
+
+```ts
+export type Slots = {
+  root: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>, HTMLElement>;
+  slot: ObjectShorthandProps<SlotProps>;
+};
+export interface Props extends ComponentProps<Slots> {
+  /* ... */
+}
+export interface State extends ComponentState<Slots> {
+  /* ... */
+}
+
+// use*State hook
+function useState({slot, ...props}: Props): State {
+  // ...
+  return {
+    slot: resolveShorthand(slot)
+    root: getNativeElementProps(state.components.root, props),
+  };
+}
+```
+
+### Make _slots_ support `as` property (but with restrictions)
```

Comment:
If 'as' is added per-slot, then is `components` a duplicate way of specifying the component type? E.g. are these two the same thing?
```jsx
<Component label={{ as: 'span', children: 'foo' }} />
<Component components={{ label: 'span' }} label="foo" /> 
```

Reply:
I've been thinking about an alternative that would allow an optional tuple of [component, shorthandProps] for a slot, instead of the `components` prop. E.g., the equivalent to above would be:
```tsx
<Component label={[ 'span', 'foo' ]} />
```
The benefit being that it still supports forwarding the `as` prop to a custom component:
```tsx
<Component label={[ MyLabelWithAsProp, { as: 'span', children: 'foo' } ]} />
```

I wonder if that would be a possible alternative to supporting `as` natively in all slots?

Reply:
That would conflict with possible future RFC ideas like the one proposed by @smhigley https://github.com/microsoft/fluentui/pull/19266

Reply:
> If 'as' is added per-slot, then is `components` a duplicate way of specifying the component type? E.g. are these two the same thing?
> 
> ```js
> <Component label={{ as: 'span', children: 'foo' }} />
> <Component components={{ label: 'span' }} label="foo" /> 
> ```

well, it would be... but `components` is also not available in the properties, it's state only

Reply:
It's also good to make it clear the main difference between `as` and `component` is what they support. `as` is limited to be a mechanism for semantic elements (change `button` to `div`), while `component` doesn't have that limitation, and that's the main reason why `as` is explicit and `components` isn't. The repercussion of making `components` somehow so accessible by the user would lean towards some undesired unmapped behaviors

Reply:
> That would conflict with possible future RFC ideas like the one proposed by @smhigley #19266

My idea was actually an attempt to solve a problem in that RFC--that you couldn't specify a different component per-slot--and would end up working nicely with her proposal. That said, Shift had some detailed comments on her RFC that might change the direction there (and obviate the problem I was trying to solve).

> well, it would be... but `components` is also not available in the properties, it's state only

Ah, I misunderstood from #18642 -- I thought `components` was being exposed as a prop. Thanks for clarifying 👍